src/core/TESTING.md

```diff
@@ -0,0 +1,210 @@
+# Testing Kibana Plugins
+
+This document outlines best practices and patterns for testing Kibana Plugins.
+
+- [Strategy](#strategy)
+- [Core Integrations](#core-integrations)
+  - [Core Mocks](#core-mocks)
+  - [Strategies for specific Core APIs](#strategies-for-specific-core-apis)
+      - [HTTP Routes](#http-routes)
+      - [SavedObjects](#savedobjects)
+      - [Elasticsearch](#elasticsearch)
+- [Plugin Integrations](#plugin-integrations)
+- [Plugin Contracts](#plugin-contracts)
+
+## Strategy
+
+In general, we recommend three tiers of tests:
+- Unit tests: small, fast, exhaustive, make heavy use of mocks for external dependencies 
+- Integration tests: higher-level tests that verify API behavior through sending real HTTP requests to Kibana server
+  - **TODO: what's the equivalent of integration tests for frontend code? Karma replacement?**
+- Functional tests: full end-to-end tests that verify user-facing behavior through the browser
+
+These tiers should roughly follow the traditional "testing pyramid", where there are more exhaustive testing at the unit level, fewer at the integration level, and very few at the functional level. 
+
+## New concerns in the Kibana Platform
+
+The Kibana Platform introduces new concepts that legacy plugins did not have concern themselves with. Namely:
+- **Lifecycles**: plugins now have explicit lifecycle methods that must interop with Core APIs and other plugins.
+- **Shared runtime**: plugins now all run in the same process at the same time. On the frontend, this is different behavior than the legacy plugins. Developers should take care not to break other plugins when interacting with their enviornment (Node.js or Browser).
+- **Single page application**: Kibana's frontend is now a single-page application where all plugins are running, but only one application is mounted at a time. Plugins need to handle mounting and unmounting, cleanup, and avoid overriding global browser behaviors in this shared space.
```

Comment:
Probably we should introduce autotests for memory leak detection. just a handful for the most critical parts https://github.com/samccone/drool and friends

Reply:
Using drool to verify app unmounting cleaning would be interesting for sure. It appears we could have a single mechanism that tests each application automatically.

Reply:
should I create an issue to investigate this option? it's not critical problem right now, can be done later